lib/services/garbageCollector.js

```diff
@@ -22,120 +22,128 @@
 
 const
-  OneHour = 3600000,
-  OneDay = OneHour * 24,
   Bluebird = require('bluebird'),
   Request = require('kuzzle-common-objects').Request;
 
-function GarbageCollector(kuzzle) {
-  this.timer = null;
+// Time constants
+const 
+  oneHour = 3600000,
+  oneDay = oneHour * 24;
+
+// There can be only one
+let _highlander = null;
+
+class GarbageCollector {
+  constructor(kuzzle) {
+    this.kuzzle = kuzzle;
+  }
+
+  init () {
+    // We do not want to trigger the garbage collector right away:
+    // if for whatever reason Kuzzle is (re)started on a production
+    // environment (after a crash, because of autoscaling, etc),
+    // the last thing a client needs is to clean up trashcans
+    setTimeout(() => this.run(), this.kuzzle.config.services.garbageCollector.cleanInterval || oneDay);
 
-  this.init = function garbageCollectorInit() {
-    this.run();
     return Bluebird.resolve(this);
-  };
+  }
 
-  this.run = function garbageCollectorRun () {
+  run () {
     let ids = [];
-    const
-      body = {
-        from: 0,
-        size: kuzzle.config.services.garbageCollector.maxDelete || 10000,
-        sort: [{ '_kuzzle_info.deletedAt': { unmapped_type: 'date'} }],
-        query: {
-          bool: {
-            should: [
-              {
-                term: {
-                  '_kuzzle_info.active': false
-                }
-              }
-            ]
-          }
-        }
-      };
-
-    if (this.timer) {
-      clearTimeout(this.timer);
-    }
 
-    if (kuzzle.funnel.overloaded) {
-      // Reschedule the task to run in 1 hour
-      this.timer = setTimeout(() => this.run(), OneHour);
+    if (this.kuzzle.funnel.overloaded) {
+      // Reschedule the task to be run in 1 hour
+      setTimeout(() => this.run(), oneHour);
 
-      return Bluebird.resolve({ids});
+      return Bluebird.resolve();
     }
 
-    // Reschedule the task to be executed every 24 hours by default
-    this.timer = setTimeout(() => this.run(), kuzzle.config.services.garbageCollector.cleanInterval || OneDay);
-
-    return kuzzle.pluginsManager.trigger('gc:start')
+    return this.kuzzle.pluginsManager.trigger('gc:start')
       .then(() => {
-        kuzzle.pluginsManager.trigger('log:info', '[GC] Started');
+        this.kuzzle.pluginsManager.trigger('log:info', '[GC] Started');
 
-        return kuzzle.services.list.storageEngine.listIndexes()
-          .then(res => {
-            const indexesPromises = [];
+        // Builds an array of {index, collection} pairs
+        const indexes = Object.keys(this.kuzzle.indexCache.indexes)
+          .filter(index => index[0] !== '%'); // skipping reserved indexes
 
-            res.indexes.forEach(index => {
-              if (index === kuzzle.internalEngine.index) {
-                return;
-              }
+        const collections = [];
+
+        for (const index of indexes) {
+          collections.push(...this.kuzzle.indexCache.indexes[index].map(collection => ({index, collection})));
```

Comment:
nitpicking: we don't have any control on the number of collections per index. I would rather use `concat` here

Reply:
fixed